compiler/rustc_typeck/src/expr_use_visitor.rs

```diff
@@ -470,14 +470,24 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
         }
     }
 
-    fn walk_local<F>(&mut self, expr: &hir::Expr<'_>, pat: &hir::Pat<'_>, mut f: F)
-    where
+    fn walk_local<F>(
+        &mut self,
+        expr: &hir::Expr<'_>,
+        pat: &hir::Pat<'_>,
+        els: Option<&hir::Block<'_>>,
+        mut f: F,
+    ) where
         F: FnMut(&mut Self),
     {
         self.walk_expr(expr);
         let expr_place = return_if_err!(self.mc.cat_expr(expr));
         f(self);
-        self.walk_irrefutable_pat(&expr_place, &pat);
+        if let Some(els) = els {
+            // borrowing because we need to test the descriminant
+            self.borrow_expr(expr, ImmBorrow);
```

Comment:
I might be wrong here. It looks necessary to borrow the initializer when it is a use of an upvar. This is one example:

https://github.com/rust-lang/rust/blob/683a9c8391f73a5fd9b39efd7b1179f6a2264ae2/compiler/rustc_typeck/src/check/_match.rs#L134

Without this, we will have ICE later in building MIR when lowering the pattern bindings. `match_pairs` are not cleared because the ancestor path check fails and the `orig_expected` upvar fails to resolve.